### PR TITLE
[Dev] Feat(moe): Gated delta net context parallel (CP)

### PR DESCRIPTION
**Pull request by yuzhongw-nvidia**
# What does this PR do ?

Support context parallel (CP) for Gated delta net (GDN) towards long context training for Qwen3-next series.

[MR for main](https://github.com/NVIDIA/Megatron-LM/pull/2642)

## Methodology

Gated Delta Net (GDN) follows Mamba context parallelization scheme, which adopts all-to-all communication to convert the context parallel for the computation of attention.

Specifically, q, k, v, gate, beta, alpha are initially distributed along the sequence dimension (shape: `[s / CP, b, head, ...]`). We use all-to-all communication to convert it to head-parallel (shape: `[s, b, head / CP, ...]`). Then, we can compute the attention parallelly on each CP rank, which has the full sequence but only partial heads. Finally, we use all-to-all communication to convert it back to context-parallel (shape: `[s / CP, b, head, ...]`).

## Usage

Using argument `--context-parallel-size`.

## Convergence

image

image

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by hxbai** [comment]
LGTM.

File: megatron/core/ssm/gated_delta_net.py [hunk 1]
```diff
@@ -271,8 +274,11 @@ def forward(
             rotary_pos_emb (Optional[Union[Tensor, Tuple[Tensor, Tensor]]]): Rotary
                 embedding tensor(s).
             rotary_pos_cos (Optional[Tensor]): Rotary embedding cosine.
+                No use for GDN, just for compatibility with Attention class.
             rotary_pos_sin (Optional[Tensor]): Rotary embedding sine.
+                No use for GDN, just for compatibility with Attention class.
             rotary_pos_cos_sin (Optional[Tensor]): Combined rotary embedding cosine and sine.
+                No use for GDN, just for compatibility with Attention class.
```
Comment (Victarry) [inline megatron/core/ssm/gated_delta_net.py]: If these args are not used, I think we could just add `**kwargs` to the forward function and remove these positional args?

**Comment by Victarry** [comment]
Hi @yuzhongw-nvidia , could you add loss curve compare with and without CP? Also could you please update the desc to explain how the CP for GDN work?

File: megatron/core/ssm/gated_delta_net.py [hunk 2]
```diff
@@ -217,7 +220,7 @@ def __init__(
             tp_group=self.pg_collection.tp,
         )
 
-        # TODO: support CP
+        self.cp = SSMContextParallel(self.pg_collection.cp)
```
Comment (Victarry) [inline megatron/core/ssm/gated_delta_net.py]: Using the name `self.cp` is too simple to be ambiguous. Could changed to names like `self.cp_handler`

File: megatron/core/ssm/ssm_context_parallel.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+    def tensor_a2a_cp2hp(
+        self,
+        tensor: torch.Tensor,
+        seq_dim: int,
+        head_dim: int,
+        split_size_or_sections: Optional[List[int]] = None,
+        undo_attention_load_balancing: bool = True,
+    ):
+        """All-to-all context parallel to hidden parallel."""
```
Comment (Victarry) [inline megatron/core/ssm/ssm_context_parallel.py]: Could you please add docs strings for each arg?

For example, the shape requirement of the input tensor is important for users to correctly use this func.

File: megatron/core/ssm/ssm_context_parallel.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    4. Sliced views of relevant trainable variables for the current context parallel rank
+    """
+
+    def __init__(self, cp_group: torch.distributed.ProcessGroup):
+        self.cp_group = cp_group
+        self.cp_size = cp_group.size()
+        self.cp_rank = cp_group.rank()
+
+    def get_parameter_local_cp(
+        self, param: torch.Tensor, dim: int, split_size_or_sections: Optional[List[int]] = None
```
Comment (Victarry) [inline megatron/core/ssm/ssm_context_parallel.py]: According the API https://docs.pytorch.org/docs/stable/generated/torch.split.html, I think 
`split_size_or_sections` means `split_size` when int is passed. And means `split_sections` when List[int] is passed.

Based on only `List[int]` is passed here, I suggest to change its name to `split_sections` for simplicity.

File: megatron/core/ssm/ssm_context_parallel.py (hunk 3 above)
Comment (yuzhongw-nvidia) [reply megatron/core/ssm/ssm_context_parallel.py]: Done, does it look good now?

File: megatron/core/ssm/ssm_context_parallel.py (hunk 4 above)
Comment (yuzhongw-nvidia) [reply megatron/core/ssm/ssm_context_parallel.py]: Resolved. Thanks.

File: megatron/core/ssm/gated_delta_net.py (hunk 1 above)
Comment (yuzhongw-nvidia) [reply megatron/core/ssm/gated_delta_net.py]: Removing them will break the API check. Let's keep them for now.

Comment (Victarry) [reply megatron/core/ssm/gated_delta_net.py]: API check is disabled for now, I think this could be removed?

**Review by Victarry — APPROVED** [review]
LGTM
